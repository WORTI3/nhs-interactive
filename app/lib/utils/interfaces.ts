import { SearchQuery } from "@/types";

export interface SearchProps {
  onSearch: (searchResult: SearchQuery) => void;
  onResultUpdate: (answer: string) => void;
  onDone: (done: boolean) => void;
}

export interface ResultProps {
  searchQuery: SearchQuery;
  answer: string;
  done: boolean;
  onReset: () => void;
}