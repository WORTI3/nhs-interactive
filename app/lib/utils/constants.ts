export const LOCAL_KEY_ID = "interactive-local-key";
export const UNRELATED_ANSWER = "not-answerable";
export const LOCAL_DB_URL = "http://127.0.0.1:8090";

export enum OpenAIModel {
  DAVINCI_TURBO = "gpt-3.5-turbo",
};