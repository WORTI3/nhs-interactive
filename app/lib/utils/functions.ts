import { createParser, ParsedEvent, ReconnectInterval } from "eventsource-parser";

/**
 * @async
 * @function openAIStream
 * @param {string} model - The model name to use for the OpenAI API request.
 * @param {string} prompt - The prompt to send for the chat completion.
 * @param {string} apiKey - The OpenAI API key for the request.
 * @returns {Promise<ReadableStream>} A readable stream that fetches and parses the response from the OpenAI API.
 */
export const openAIStream = async (model: string, prompt: string, apiKey: string) => {
  const encoder = new TextEncoder();
  const decoder = new TextDecoder();

  const res = await fetch("https://api.openai.com/v1/chat/completions", {
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${apiKey}`
    },
    method: "POST",
    body: JSON.stringify({
      model: model,
      messages: [
        { role: "system", content: "You are a medical professional that accurately answers the user's queries based on the given text and you only get your information and facts from NHS websites." },
        { role: "user", content: prompt }
      ],
      max_tokens: 120,
      temperature: 0.0,
      stream: true
    })
  });

  if (res.status !== 200) {
    throw new Error("OpenAI API returned an error");
  }

  const stream = new ReadableStream({
    async start(controller) {
      const onParse = (event: ParsedEvent | ReconnectInterval) => {
        if (event.type === "event") {
          const data = event.data;

          if (data === "[DONE]") {
            controller.close();
            return;
          }

          try {
            const json = JSON.parse(data);
            const text = json.choices[0].delta.content;
            const queue = encoder.encode(text);
            controller.enqueue(queue);
          } catch (e) {
            controller.error(e);
          }
        }
      };

      const parser = createParser(onParse);
      for await (const chunk of res.body as any) {
        parser.feed(decoder.decode(chunk));
      }
    }
  });

  return stream;
};


/**
 * This code exports a function named `cleanSourceText` that takes a string as input and returns a modified version of the
 * string with certain characters and patterns replaced or removed. The modifications include removing leading and trailing
 * whitespace, replacing multiple consecutive newlines with a single space, replacing multiple consecutive spaces with a
 * single space, removing tabs, and removing excess newlines.
 * 
 * @function
 * @name cleanSourceText
 * @kind variable
 * @param {string} text
 * @returns {string}
 * @exports
 */
export const cleanSourceText = (text: string) => {
  return text
    .trim()
    .replace(/(\n){4,}/g, "\n\n\n")
    .replace(/\n\n/g, " ")
    .replace(/ {3,}/g, "  ")
    .replace(/\t/g, "")
    .replace(/\n+(\s*\n)*/g, "\n");
};
