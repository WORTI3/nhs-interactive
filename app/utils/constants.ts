export const LOCAL_KEY_ID = "interactive-local-key";
export const UNRELATED_ANSWER = "not-answerable";
export const LOCAL_DB_URL = "http://127.0.0.1:8090";
export const SERVICE_NAME = "NHS Interactive";
export const SERVICE_DESC = "AI powered search.";
export const SERVICE_URL = "https://nhs-interactive.vercel.app";
export const LOCALES = ['en', 'cy', 'gd'];
export const USE_AI_RESPONSE_KEY = "use-ai-response";

export enum OpenAIModel {
  DAVINCI_TURBO = "gpt-3.5-turbo",
};