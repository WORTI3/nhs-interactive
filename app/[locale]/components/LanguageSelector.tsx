import { Fragment } from 'react'
import { Menu, Transition } from '@headlessui/react'
import { IconWorld } from '@tabler/icons-react';
import Link from "next-intl/link";
import { LOCALES } from '../../utils/constants';

function classNames(...classes: any[]) {
  return classes.filter(Boolean).join(' ')
}

export default function LanguageSelector({translate, locale}: {translate: any, locale: string}) {
  return (
    <Menu as='div' className='relative max-lg:absolute max-lg:right-4 float-right inline-block text-left'>
      <div>
        {/* Hover style made need changed to match NHS */}
        <Menu.Button id='language-button' className='inline-flex w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50'>
          {translate('button.language')}
          <IconWorld
            className='-mr-1 h-5 w-5 text-gray-400'
            aria-hidden='true'
          />
        </Menu.Button>
      </div>

      <Transition
        as={Fragment}
        enter='transition ease-out duration-100'
        enterFrom='transform opacity-0 scale-95'
        enterTo='transform opacity-100 scale-100'
        leave='transition ease-in duration-75'
        leaveFrom='transform opacity-100 scale-100'
        leaveTo='transform opacity-0 scale-95'
      >
        <Menu.Items className='absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none'>
          <div className='langs'>
            {LOCALES.map(lang => {
              if (lang !== locale) {
                return (
                  <Menu.Item key={lang} data-testid="language-item">
                    {({ active }) => (
                      <Link
                        href='/'
                        locale={lang}
                        className={classNames(
                          active ? "bg-gray-100 text-gray-900" : "text-gray-700",
                          "block px-4 py-2 text-sm rounded-md"
                        )}
                      >
                        <p className='mb-0 py-2'>
                          <span
                            className={`mr-4 h-5 w-5 rounded-md fi fis fi-${lang === 'en' ? 'gb-eng' : lang === 'cy' ? 'gb-wls' : 'gb-sct'}`}
                            aria-hidden='true'
                          />
                          {lang === 'en' ? 'English' : lang === 'cy' ? 'Cymraeg' : 'Gaelic'}
                        </p>
                      </Link>
                    )}
                  </Menu.Item>
                );
              }
              return null;
            })}
          </div>
        </Menu.Items>
      </Transition>
    </Menu>
  );
}
