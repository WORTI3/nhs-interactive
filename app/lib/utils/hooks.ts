import { LogData } from '@/types';
import { useEffect, useState } from 'react';
import { LOCAL_KEY_ID, UNRELATED_ANSWER } from './constants';

/**
 * A custom React hook that takes in two parameters: `logData` of type `LogData` and `done` of
 * type `boolean`. It sets up an effect that sends a POST request to the `/api/log` endpoint with the `logData` object as
 * the request body when `done` is true or when `logData.answer` is equal to `UNRELATED_ANSWER`. The hook also manages the
 * state of `logSubmitted` using the `useState` hook to ensure that the POST request is only sent once.
 * 
 * @function
 * @name useApiLog
 * @kind variable
 * @param {LogData} logData
 * @param {boolean} done
 * @returns {void}
 * @exports
 */
export const useApiLog = (logData: LogData, done: boolean) => {
  const [logSubmitted, setLogSubmitted] = useState(false);
  useEffect(() => {
    const fetchApiLog = async () => {
      try {
        const response = await fetch('/api/log', {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(logData),
        });

        if (!response.ok) {
          // Handle error case
        }
      } catch (error) {
        // Handle network error
      }
    };
    if (!logSubmitted) {
      if (done || logData.answer === UNRELATED_ANSWER) {
        fetchApiLog();
        setLogSubmitted(true);
      };
    };
    
  }, [logData, done, logSubmitted]);
};

/**
 * A custom React hook that takes a function `setApiKey` as a parameter. It retrieves an API
 * key from local storage and sets it using the `setApiKey` function. 
 * This hook is used to manage the API key across several pages within the application.
 * 
 * @function
 * @name useApiKey
 * @kind variable
 * @param {(apiKey: string) => void} setApiKey
 * @returns {void}
 * @exports
 */
export const useApiKey = (setApiKey: (apiKey: string) => void) => {
  useEffect(() => {
    const API_KEY = localStorage.getItem(LOCAL_KEY_ID);
    if (API_KEY) {
      setApiKey(JSON.parse(API_KEY));
    }
  }, [setApiKey]);
};
