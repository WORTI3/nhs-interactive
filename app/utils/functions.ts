import { HealthAPIResponse, SearchQuery } from "@/types";
import { UNRELATED_ANSWER } from "./constants";
import { notFound } from "next/navigation";

/**
 * The `fetchAnswer` function is an asynchronous function that sends a POST request to the `/api/answer` endpoint with the
 * `query` and `lang` parameters as the request body. It then reads the response body as a stream and updates the answer
 * using the `onAnswerUpdate` callback function. It also calls the `onSearch` callback function to provide information
 * about the search query, and the `onDone` callback function to indicate that the process is done. If an error occurs, it
 * updates the answer with the `UNRELATED_ANSWER` constant.
 * 
 * @async
 * @function
 * @name fetchAnswer
 * @kind variable
 * @param {string} query
 * @param {(answer: string) => void} onAnswerUpdate
 * @param {(searchQuery: SearchQuery) => void} onSearch
 * @param {(done: boolean) => void} onDone
 * @param {string} lang
 * @returns {Promise<void>}
 * @exports
 */
export const fetchAnswer = async (query: string, onAnswerUpdate: (answer: string) => void, onSearch: (searchQuery: SearchQuery) => void, onDone: (done: boolean) => void, lang: string) => {
  try {
    const response = await fetch("/api/answer", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ query, lang }),
    });

    if (!response.ok) {
      throw new Error(response.statusText);
    }

    let { data }: {data: HealthAPIResponse} = await response.json();
    // todo: validate the response more
    // loose matching:
    let answer = data.answer;
    const aiErrorPhrases = ["i'm sorry", "i cannot provide", "i don't know"];
    if (answer.length <= 15 || aiErrorPhrases.some(phrase => answer.toLocaleLowerCase().includes(phrase))) {
      answer = UNRELATED_ANSWER;
    }
    // call this here to ensure query is shown before answer
    onSearch({ query, sourceLinks: data.sources, sourceHeadings: data.headings });
    onAnswerUpdate(answer);
    onDone(true);
  } catch (error: any) {
    onAnswerUpdate(UNRELATED_ANSWER);
    onSearch({ query: error.message ?? 'Error', sourceLinks: [], sourceHeadings: [] });
  }
};

export const checkHealthAPIStatus = async (url: string) => {
  let maintenanceMode = false;
  try {
    // monitor this to see if this will slow down api or app at all
    const res = await fetch(url, { next: { revalidate: false } });
    const data = await res.json();
    if (data.api_version !== process.env.API_VERSION || res.status !== 200){
      maintenanceMode = true;
      console.warn(JSON.stringify(data));
    }
  } catch {
    maintenanceMode = true;
  }
  return maintenanceMode;
};

export const getInternationalisation = async (locale: string) => {
  try {
    return (await import(`../../locales/${locale}.json`)).default;
  } catch (error) {
    return notFound();
  }
};

export function classNames(...classes: any[]) {
  return classes.filter(Boolean).join(' ')
}
