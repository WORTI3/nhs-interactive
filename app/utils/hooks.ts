import { LogData } from '@/types';
import { useEffect, useRef } from 'react';
import { UNRELATED_ANSWER } from './constants';

/**
 * A custom React hook that takes in two parameters: `logData` of type `LogData` and `done` of
 * type `boolean`. It sets up an effect that sends a POST request to the `/api/log` endpoint with the `logData` object as
 * the request body when `done` is true or when `logData.answer` is equal to `UNRELATED_ANSWER`. The hook also manages the
 * state of `logSubmitted` using the `useState` hook to ensure that the POST request is only sent once.
 * 
 * @function
 * @name useApiLog
 * @kind variable
 * @param {LogData} logData
 * @param {boolean} done
 * @returns {void}
 * @exports
 */
export const useApiLog = (logData: LogData, done: boolean, id: string) => {
  const isEffectActive = useRef<boolean>(false);
  useEffect(() => {
    if (!isEffectActive.current && (done || logData.answer === UNRELATED_ANSWER)) {
      const data = {
        ...logData,
        id
      }
      const fetchApiLog = async () => {
        await fetch('/api/log', {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(data),
        });
        // do we really need error handling here..?
      };

      fetchApiLog()
      isEffectActive.current = true;
    }
  }, [logData, done]);
};
