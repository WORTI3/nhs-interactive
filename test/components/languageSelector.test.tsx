import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import LanguageSelector from '../../app/[locale]/components/LanguageSelector';
import { load } from 'cheerio';
import { LOCALES } from '../../app/utils/constants';
import { NextIntlClientProvider } from 'next-intl';
import { act } from 'react';

describe('LanguageSelector component', () => {
  let translation: { 'button.language': string };
  const mockTranslate = jest.fn((key: keyof { 'button.language': string }) => {
    const translations = translation;
    return translations[key];
  });

  beforeEach(() => {
    jest.clearAllMocks();
    Object.defineProperty(window, 'ResizeObserver', {
      writable: true,
      value: jest.fn().mockImplementation(() => ({
        observe: jest.fn(),
        unobserve: jest.fn(),
        disconnect: jest.fn(),
      })),
    });
  });

  function toRender(messages: any, locale: string) {
    translation = { 'button.language': messages.header.button.language };
    return (<NextIntlClientProvider messages={messages} locale={locale}>
      <LanguageSelector locale={locale} translate={mockTranslate} />
    </NextIntlClientProvider>)
  };

  test.each(LOCALES)(
    'rendered component should match snapshot (%s)',
    (locale) => {
      const messages = require(`../../locales/${locale}.json`);
      const { container } = render(toRender(messages, locale));
      expect(container).toMatchSnapshot();
    }
  );

  test.each(LOCALES)(
    'should render the component with the correct links & elements (%s)',
    (locale) => {
      const messages = require(`../../locales/${locale}.json`);
      const { container } = render(toRender(messages, locale));
      const html = container.innerHTML;
      const $ = load(html);
      
      const languageButton = $('button');
      
      expect(languageButton.length).toBe(1);
      expect(languageButton.text()).toStrictEqual(messages.header.button.language);
    }
  );

  test.each(LOCALES)(
    'should render the component with the correct links & elements (%s)',
    async (locale) => {
      const messages = require(`../../locales/${locale}.json`);
      const { container } = render(toRender(messages, locale));
      const html = container.innerHTML;
      
      const submitButton = screen.getByRole("button", { name: messages.header.button.language });
      act(() => {
        fireEvent.click(submitButton);
      });
      expect(container).toMatchSnapshot();

      const languageItems = screen.getAllByTestId('language-item');
      const expectedLanguages = LOCALES.filter((lang) => lang !== locale);
      const expectedLanguagesText = expectedLanguages.map((lang) =>
        lang === 'en' ? 'English' : lang === 'cy' ? 'Cymraeg' : 'Gaelic'
      );

      expect(languageItems).toHaveLength(expectedLanguages.length);
      languageItems.forEach((item, index) => {
        const languageText = expectedLanguagesText[index];
        expect(item).toHaveTextContent(languageText);
      });
    }
  );
});
