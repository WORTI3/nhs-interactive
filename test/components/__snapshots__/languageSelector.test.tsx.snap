// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelector component rendered component should match snapshot (cy) 1`] = `
<div>
  <div
    class="absolute right-5 top-5 max-xl:right-5 sm:right-8 text-left"
    data-headlessui-state=""
  >
    <div>
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        class="inline-flex focus:outline-black focus:outline-8 focus:ring-4 focus:ring-yellow-400 w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm hover:bg-gray-50"
        data-headlessui-state=""
        id="language-button"
        type="button"
      >
        Iaith
        <svg
          aria-hidden="true"
          class="tabler-icon tabler-icon-world -mr-1 h-5 w-5 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
          />
          <path
            d="M3.6 9h16.8"
          />
          <path
            d="M3.6 15h16.8"
          />
          <path
            d="M11.5 3a17 17 0 0 0 0 18"
          />
          <path
            d="M12.5 3a17 17 0 0 1 0 18"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`LanguageSelector component rendered component should match snapshot (en) 1`] = `
<div>
  <div
    class="absolute right-5 top-5 max-xl:right-5 sm:right-8 text-left"
    data-headlessui-state=""
  >
    <div>
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        class="inline-flex focus:outline-black focus:outline-8 focus:ring-4 focus:ring-yellow-400 w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm hover:bg-gray-50"
        data-headlessui-state=""
        id="language-button"
        type="button"
      >
        Language
        <svg
          aria-hidden="true"
          class="tabler-icon tabler-icon-world -mr-1 h-5 w-5 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
          />
          <path
            d="M3.6 9h16.8"
          />
          <path
            d="M3.6 15h16.8"
          />
          <path
            d="M11.5 3a17 17 0 0 0 0 18"
          />
          <path
            d="M12.5 3a17 17 0 0 1 0 18"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`LanguageSelector component rendered component should match snapshot (gd) 1`] = `
<div>
  <div
    class="absolute right-5 top-5 max-xl:right-5 sm:right-8 text-left"
    data-headlessui-state=""
  >
    <div>
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        class="inline-flex focus:outline-black focus:outline-8 focus:ring-4 focus:ring-yellow-400 w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm hover:bg-gray-50"
        data-headlessui-state=""
        id="language-button"
        type="button"
      >
        CÃ nan
        <svg
          aria-hidden="true"
          class="tabler-icon tabler-icon-world -mr-1 h-5 w-5 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
          />
          <path
            d="M3.6 9h16.8"
          />
          <path
            d="M3.6 15h16.8"
          />
          <path
            d="M11.5 3a17 17 0 0 0 0 18"
          />
          <path
            d="M12.5 3a17 17 0 0 1 0 18"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`LanguageSelector component should render the component with the correct links & elements (cy) 1`] = `
<div>
  <div
    class="absolute right-5 top-5 max-xl:right-5 sm:right-8 text-left"
    data-headlessui-state="open"
    data-open=""
  >
    <div>
      <button
        aria-controls="headlessui-menu-items-:rj:"
        aria-expanded="true"
        aria-haspopup="menu"
        class="inline-flex focus:outline-black focus:outline-8 focus:ring-4 focus:ring-yellow-400 w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm hover:bg-gray-50"
        data-active=""
        data-headlessui-state="open active"
        data-open=""
        id="language-button"
        type="button"
      >
        Iaith
        <svg
          aria-hidden="true"
          class="tabler-icon tabler-icon-world -mr-1 h-5 w-5 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
          />
          <path
            d="M3.6 9h16.8"
          />
          <path
            d="M3.6 15h16.8"
          />
          <path
            d="M11.5 3a17 17 0 0 0 0 18"
          />
          <path
            d="M12.5 3a17 17 0 0 1 0 18"
          />
        </svg>
      </button>
    </div>
    <div
      aria-labelledby="language-button"
      class="absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none transition ease-out duration-100 transform opacity-0 scale-95"
      data-headlessui-state="open"
      data-open=""
      id="headlessui-menu-items-:rj:"
      role="menu"
      style="--button-width: 0px;"
      tabindex="0"
    >
      <div
        class="langs"
        role="none"
      >
        <a
          class="text-gray-700 block px-4 py-2 text-sm rounded-md"
          data-headlessui-state=""
          data-testid="language-item"
          href="/en"
          id="headlessui-menu-item-:rk:"
          role="menuitem"
          tabindex="-1"
        >
          <p
            class="mb-0 py-2"
          >
            <span
              aria-hidden="true"
              class="mr-4 h-5 w-5 rounded-md fi fis fi-gb-eng"
            />
            English
          </p>
        </a>
        <a
          class="text-gray-700 block px-4 py-2 text-sm rounded-md"
          data-headlessui-state=""
          data-testid="language-item"
          href="/gd"
          id="headlessui-menu-item-:rl:"
          role="menuitem"
          tabindex="-1"
        >
          <p
            class="mb-0 py-2"
          >
            <span
              aria-hidden="true"
              class="mr-4 h-5 w-5 rounded-md fi fis fi-gb-sct"
            />
            Gaelic
          </p>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`LanguageSelector component should render the component with the correct links & elements (en) 1`] = `
<div>
  <div
    class="absolute right-5 top-5 max-xl:right-5 sm:right-8 text-left"
    data-headlessui-state="open"
    data-open=""
  >
    <div>
      <button
        aria-controls="headlessui-menu-items-:re:"
        aria-expanded="true"
        aria-haspopup="menu"
        class="inline-flex focus:outline-black focus:outline-8 focus:ring-4 focus:ring-yellow-400 w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm hover:bg-gray-50"
        data-active=""
        data-headlessui-state="open active"
        data-open=""
        id="language-button"
        type="button"
      >
        Language
        <svg
          aria-hidden="true"
          class="tabler-icon tabler-icon-world -mr-1 h-5 w-5 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
          />
          <path
            d="M3.6 9h16.8"
          />
          <path
            d="M3.6 15h16.8"
          />
          <path
            d="M11.5 3a17 17 0 0 0 0 18"
          />
          <path
            d="M12.5 3a17 17 0 0 1 0 18"
          />
        </svg>
      </button>
    </div>
    <div
      aria-labelledby="language-button"
      class="absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none transition ease-out duration-100 transform opacity-0 scale-95"
      data-headlessui-state="open"
      data-open=""
      id="headlessui-menu-items-:re:"
      role="menu"
      style="--button-width: 0px;"
      tabindex="0"
    >
      <div
        class="langs"
        role="none"
      >
        <a
          class="text-gray-700 block px-4 py-2 text-sm rounded-md"
          data-headlessui-state=""
          data-testid="language-item"
          href="/cy"
          id="headlessui-menu-item-:rf:"
          role="menuitem"
          tabindex="-1"
        >
          <p
            class="mb-0 py-2"
          >
            <span
              aria-hidden="true"
              class="mr-4 h-5 w-5 rounded-md fi fis fi-gb-wls"
            />
            Cymraeg
          </p>
        </a>
        <a
          class="text-gray-700 block px-4 py-2 text-sm rounded-md"
          data-headlessui-state=""
          data-testid="language-item"
          href="/gd"
          id="headlessui-menu-item-:rg:"
          role="menuitem"
          tabindex="-1"
        >
          <p
            class="mb-0 py-2"
          >
            <span
              aria-hidden="true"
              class="mr-4 h-5 w-5 rounded-md fi fis fi-gb-sct"
            />
            Gaelic
          </p>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`LanguageSelector component should render the component with the correct links & elements (gd) 1`] = `
<div>
  <div
    class="absolute right-5 top-5 max-xl:right-5 sm:right-8 text-left"
    data-headlessui-state="open"
    data-open=""
  >
    <div>
      <button
        aria-controls="headlessui-menu-items-:ro:"
        aria-expanded="true"
        aria-haspopup="menu"
        class="inline-flex focus:outline-black focus:outline-8 focus:ring-4 focus:ring-yellow-400 w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm hover:bg-gray-50"
        data-active=""
        data-headlessui-state="open active"
        data-open=""
        id="language-button"
        type="button"
      >
        CÃ nan
        <svg
          aria-hidden="true"
          class="tabler-icon tabler-icon-world -mr-1 h-5 w-5 text-gray-400"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
          />
          <path
            d="M3.6 9h16.8"
          />
          <path
            d="M3.6 15h16.8"
          />
          <path
            d="M11.5 3a17 17 0 0 0 0 18"
          />
          <path
            d="M12.5 3a17 17 0 0 1 0 18"
          />
        </svg>
      </button>
    </div>
    <div
      aria-labelledby="language-button"
      class="absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none transition ease-out duration-100 transform opacity-0 scale-95"
      data-headlessui-state="open"
      data-open=""
      id="headlessui-menu-items-:ro:"
      role="menu"
      style="--button-width: 0px;"
      tabindex="0"
    >
      <div
        class="langs"
        role="none"
      >
        <a
          class="text-gray-700 block px-4 py-2 text-sm rounded-md"
          data-headlessui-state=""
          data-testid="language-item"
          href="/en"
          id="headlessui-menu-item-:rp:"
          role="menuitem"
          tabindex="-1"
        >
          <p
            class="mb-0 py-2"
          >
            <span
              aria-hidden="true"
              class="mr-4 h-5 w-5 rounded-md fi fis fi-gb-eng"
            />
            English
          </p>
        </a>
        <a
          class="text-gray-700 block px-4 py-2 text-sm rounded-md"
          data-headlessui-state=""
          data-testid="language-item"
          href="/cy"
          id="headlessui-menu-item-:rq:"
          role="menuitem"
          tabindex="-1"
        >
          <p
            class="mb-0 py-2"
          >
            <span
              aria-hidden="true"
              class="mr-4 h-5 w-5 rounded-md fi fis fi-gb-wls"
            />
            Cymraeg
          </p>
        </a>
      </div>
    </div>
  </div>
</div>
`;
